### DISCUSSION TITLE
Questions on FastNumericArrayConverter

### DISCUSSION THREAD

**Discussion by guohaoqiang** [NVIDIA/cutlass#1336] [Q&A]
It converts 4 uint8 to 4 half precision. I have no idea why it adds 0x66006600 here. Anyone could help with it please? Thanks

<img width="864" alt="Screen Shot 2024-02-05 at 7 52 29 PM" src="https://github.com/NVIDIA/cutlass/assets/5676786/005f0428-56cb-4249-a441-7c27745df578">

**Comment by hwu36** [comment]
there is a certain range in which a data `d`'s float hex is the same as its int hex.  then, we can add a bias to a value to push it to be within this range.  we just need to subtract the bias after the conversion to restore its previous value.

the assumption here is that add/sub is faster than the real conversion.

**Reply by guohaoqiang** [reply]
Thank you. I'm trying to figure out if 0x6600 is unique, namely whether 0x6600 is replaceable or not.
I will post my code and findings later this week.

**Reply by hwu36** [reply]
it is not unique.  any bias that can push the initial value to the range in which float and int hex are the same is okay.

**Reply by guohaoqiang** [reply]
How to get the magic number is a trick.

let's say:
int8_t a = -1; // binary 1111 1111
int16_t b = a + 0x6600; // b = 0110 0101 1111 1111

Now we have 
b(hex) = 0x65ff
b(decimal) = 26111
but b(__half) = 1535 != 26111 

https://evanw.github.io/float-toy/ 

<img width="443" alt="Screen Shot 2024-02-10 at 12 06 08 AM" src="https://github.com/NVIDIA/cutlass/assets/5676786/8130cffe-dc8b-487a-b3e8-8e1a1a693d74">

**Reply by hwu36** [reply]
0x6600 is for uint8_t, not int8_t

**Comment by hwu36** [comment]
you can choose a random int value, e.g. 1.  to go through these instructions to see how the hex change.

**Comment by hwu36** [comment]
this tool might be helpful for you to understand https://gregstoll.com/~gregstoll/floattohex/
